src/core/perf_event_array/v0_4_0.rs

```diff
@@ -0,0 +1,120 @@
+use crate::cpuonline;
+use crate::types::MutPointer;
+
+use bcc_sys::bccapi::*;
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+
+#[derive(Debug)]
+pub struct PerfEventArray {
+    name: String,
+    ev_type: u32,
+    ev_config: u32,
+    table_fd: i32,
+    cpu_fd: HashMap<usize, i32>,
+}
+
+impl PerfEventArray {
+    pub fn new(name: String, ev_type: u32, ev_config: u32, table_fd: i32) -> Self {
+        let hash_map: HashMap<usize, i32> = HashMap::new();
+        Self {
+            name,
+            ev_type,
+            ev_config,
+            table_fd,
+            cpu_fd: hash_map,
+        }
+    }
+
+    pub fn close_all_cpu(&mut self) {
+        self.cpu_fd.retain(|_, v| {
+            if *v < 0 {
+                unsafe { bpf_close_perf_event_fd(*v) };
+            }
+
+            false
+        });
+    }
+
+    #[allow(dead_code)]
+    pub fn close_on_cpu(&mut self, cpu: usize) {
+        if let Some(fd) = self.cpu_fd.remove(&cpu) {
+            unsafe { bpf_close_perf_event_fd(fd) };
+        }
+    }
+
+    pub fn open_all_cpu(&mut self) -> Result<(), String> {
+        let cpus = cpuonline::get();
+
+        if let Ok(cpus) = cpus {
+            for cpu in cpus {
+                let result = self.open_on_cpu(cpu);
+
+                if let Err(e) = result {
+                    self.close_all_cpu();
+                    return Err(e);
+                }
+            }
+        }
+
+        Ok(())
```

Comment:
Is returning Ok if cpuonline::get() fails the correct behavior? Should we maybe raise an error instead?